agent/engine/service_connect/manager.go

```diff
@@ -0,0 +1,86 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package serviceconnect
+
+import (
+	"os"
+	"path/filepath"
+
+	apicontainer "github.com/aws/amazon-ecs-agent/agent/api/container"
+	apitask "github.com/aws/amazon-ecs-agent/agent/api/task"
+
+	dockercontainer "github.com/docker/docker/api/types/container"
+)
+
+type Manager interface {
+	InitializeContainer(taskId string, container *apicontainer.Container, hostConfig *dockercontainer.HostConfig) error
+}
+
+type manager struct {
+	relayPathContainer  string
+	relayPathHost       string
+	relayFileName       string
+	relayENV            string
+	statusPathContainer string
+	statusPathHostRoot  string
+	statusFileName      string
+	statusENV           string
+}
+
+func NewManager() *manager {
+	return &manager{
+		relayPathContainer:  defaultRelayPathContainer,
+		relayPathHost:       defaultRelayPathHost,
+		relayFileName:       defaultRelayFileName,
+		relayENV:            defaultRelayENV,
+		statusPathContainer: defaultStatusPathContainer,
+		statusPathHostRoot:  defaultStatusPathHostRoot,
+		statusFileName:      defaultStatusFileName,
+		statusENV:           defaultStatusENV,
+	}
+}
+
+func IsServiceConnectEnabledTask(task *apitask.Task) bool {
+	return task.IsServiceConnectEnabled()
+}
+
+func (m *manager) InitializeTask(task *apitask.Task, hostConfig *dockercontainer.HostConfig) (rErr error) {
+	return m.initServiceConnectDirectoryMounts(task.GetID(), task.GetServiceConnectContainer(), hostConfig)
+}
+
+func getBindMountMapping(hostDir, containerDir string) string {
+	return hostDir + ":" + containerDir
+}
+
+func (m *manager) initServiceConnectDirectoryMounts(taskId string, container *apicontainer.Container, hostConfig *dockercontainer.HostConfig) (rErr error) {
+	var instancePath string
+	instancePath = filepath.Join(m.statusPathHostRoot, taskId)
+	if _, err := os.Stat(instancePath); os.IsNotExist(err) {
+		os.MkdirAll(instancePath, 0770)
+	} else if err != nil {
+		return err
+	}
+	if _, err := os.Stat(m.relayPathHost); os.IsNotExist(err) {
+		os.MkdirAll(m.relayPathHost, 0770)
+	} else if err != nil {
+		return err
+	}
+
+	hostConfig.Binds = append(hostConfig.Binds, getBindMountMapping(instancePath, m.statusPathContainer))
+	hostConfig.Binds = append(hostConfig.Binds, getBindMountMapping(m.relayPathHost, m.relayPathContainer))
+
+	container.Environment[m.relayENV] = filepath.Join(m.relayPathContainer, m.relayFileName)
```

Comment:
can we use `container.MergeEnvironmentVariables`. The reason is `Environment` is not guaranteed to be initialized and `container.MergeEnvironmentVariables` does so if needed.

Reply:
updated